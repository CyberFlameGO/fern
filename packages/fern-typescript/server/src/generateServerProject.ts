import { IntermediateRepresentation } from "@fern-fern/ir-model";
import {
    createDirectoryAndExportFromModule,
    DependencyManager,
    GeneratedProjectSrcInfo,
    generateTypeScriptProject,
} from "@fern-typescript/commons";
import { generateModelFiles } from "@fern-typescript/model";
import { Volume } from "memfs/lib/volume";
import { Directory } from "ts-morph";
import { generateHttpService } from "./http/generateHttpService";

export async function generateServerProject({
    intermediateRepresentation,
    packageName,
    packageVersion,
    volume,
}: {
    intermediateRepresentation: IntermediateRepresentation;
    packageName: string;
    packageVersion: string | undefined;
    volume: Volume;
}): Promise<void> {
    await generateTypeScriptProject({
        volume,
        packageName,
        packageVersion,
        generateSrc: (directory) => generateServerFiles({ intermediateRepresentation, directory }),
    });
}

async function generateServerFiles({
    intermediateRepresentation,
    directory,
}: {
    intermediateRepresentation: IntermediateRepresentation;
    directory: Directory;
}): Promise<GeneratedProjectSrcInfo> {
    const dependencyManager = new DependencyManager();

    const modelContext = generateModelFiles({
        modelDirectory: createDirectoryAndExportFromModule(directory, "model"),
        intermediateRepresentation,
        dependencyManager,
        mode: "server",
    });

    const servicesDirectory = createDirectoryAndExportFromModule(directory, "services");
    for (const service of intermediateRepresentation.services.http) {
        await generateHttpService({
            service,
            servicesDirectory,
            modelContext,
            dependencyManager,
        });
    }

    return {
        dependencies: dependencyManager.getDependencies(),
    };
}
