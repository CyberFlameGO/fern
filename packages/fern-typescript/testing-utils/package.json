{
  "name": "@fern-typescript/testing-utils",
  "version": "0.0.0",
  "repository": {
    "type": "git",
    "url": "git@github.com:fern-api/fern.git",
    "directory": "packages/fern-typescript/testing-utils"
  },
  "files": [
    "lib"
  ],
  "type": "module",
  "source": "src/index.ts",
  "module": "src/index.ts",
  "main": "src/index.ts",
  "sideEffects": false,
  "scripts": {
    "clean": "rm -rf ./lib && tsc --build --clean",
    "compile": "tsc --build",
    "test": "yarn run compile && jest --passWithNoTests",
    "lint:eslint": "eslint --max-warnings 0 . --ignore-path=../../../.eslintignore",
    "lint:eslint:fix": "eslint --max-warnings 0 . --ignore-path=../../../.eslintignore --fix",
    "format": "prettier --write --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "format:check": "prettier --check --ignore-unknown --ignore-path ../../../shared/.prettierignore \"**\"",
    "depcheck": "depcheck"
  },
  "dependencies": {
    "@fern-api/ir-generator": "workspace:*",
    "@fern-api/workspace-loader": "workspace:*",
    "@fern-fern/ir-model": "0.0.49",
    "@fern-typescript/commons": "workspace:*",
    "execa": "^5.1.1",
    "is-ci": "^3.0.1",
    "memfs": "^3.4.1",
    "typescript": "4.6.4"
  },
  "devDependencies": {
    "@babel/core": "^7.18.9",
    "@babel/preset-env": "^7.18.9",
    "@babel/preset-typescript": "^7.18.6",
    "@types/is-ci": "^3.0.0",
    "@types/jest": "^28.1.6",
    "@types/node": "^18.6.1",
    "depcheck": "^1.4.3",
    "eslint": "^8.20.0",
    "jest": "^28.1.3",
    "prettier": "^2.7.1",
    "ts-jest": "^28.0.7"
  }
}
