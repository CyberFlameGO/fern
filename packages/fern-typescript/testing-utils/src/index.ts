export { itFernETE } from "./itFernETE";
export { installAndCompileGeneratedProjects } from "./runEteTest";
