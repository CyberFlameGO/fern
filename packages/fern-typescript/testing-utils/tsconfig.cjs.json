{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib/cjs", "rootDir": "src", "module": "CommonJS" },
  "include": ["./src"],
  "references": [
    { "path": "../../cli/tsconfig.cjs.json" },
    { "path": "../../ir/ir-generator/tsconfig.cjs.json" },
    { "path": "../../workspace-parser/tsconfig.cjs.json" },
    { "path": "../commons/tsconfig.cjs.json" }
  ]
}
