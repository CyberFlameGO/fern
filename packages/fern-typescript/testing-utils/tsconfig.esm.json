{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib/esm", "rootDir": "src", "module": "esnext" },
  "include": ["./src"],
  "references": [
    { "path": "../../cli/tsconfig.esm.json" },
    { "path": "../../compiler/compiler/tsconfig.esm.json" },
    { "path": "../commons/tsconfig.esm.json" }
  ]
}
