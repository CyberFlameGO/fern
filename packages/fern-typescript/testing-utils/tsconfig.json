{
  "extends": "../../../shared/tsconfig.shared.json",
  "compilerOptions": { "composite": true, "outDir": "lib", "rootDir": "src" },
  "include": ["./src"],
  "references": [
    { "path": "../../generation/ir/ir-generator" },
    { "path": "../../workspace-loader" },
    { "path": "../commons" }
  ]
}
