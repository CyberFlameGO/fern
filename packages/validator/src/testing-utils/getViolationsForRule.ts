import { parseWorkspaceDefinition } from "@fern-api/workspace-parser";
import { visitFernYamlAst } from "@fern-api/yaml-schema";
import { createAstVisitorForRules } from "../createAstVisitorForRules";
import { Rule, RuleViolation } from "../Rule";

export declare namespace getViolationsForRule {
    export interface Args {
        rule: Rule;
        absolutePathToDefinition: string;
    }
}

export async function getViolationsForRule({
    rule,
    absolutePathToDefinition,
}: getViolationsForRule.Args): Promise<RuleViolation[]> {
    const parseResult = await parseWorkspaceDefinition({
        name: undefined,
        absolutePathToDefinition,
    });
    if (!parseResult.didSucceed) {
        throw new Error("Failed to parse workspace");
    }

    const ruleRunner = await rule.create({ workspace: parseResult.workspace });
    const violations: RuleViolation[] = [];

    for (const [relativeFilePath, contents] of Object.entries(parseResult.workspace.files)) {
        const visitor = createAstVisitorForRules({
            relativeFilePath,
            contents,
            ruleRunners: [ruleRunner],
            addViolations: (newViolations) => {
                violations.push(...newViolations);
            },
        });
        await visitFernYamlAst(contents, visitor);
    }

    return violations;
}
