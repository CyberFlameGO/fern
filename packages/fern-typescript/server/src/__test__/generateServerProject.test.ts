import { itFernETE } from "@fern-typescript/testing-utils";
import path from "path";
import { generateServerProject } from "../generateServerProject";

const FIXTURES_DIR = "fixtures";
const FIXTURES = ["posts"];

const MOCK_HELPERS_MANAGERS = new HelperManager({ encodings: {} });

describe("generateServerProject", () => {
    for (const fixture of FIXTURES) {
        itFernETE(fixture, {
            testFile: __filename,
            pathToFixture: path.join(FIXTURES_DIR, fixture),
            generateFiles: async ({ volume, intermediateRepresentation }) => {
                await generateServerProject({
                    packageName: fixture,
                    packageVersion: "0.0.0",
                    volume,
                    intermediateRepresentation,
                    helperManager: MOCK_HELPERS_MANAGERS,
                });
            },
        });
    }
});
