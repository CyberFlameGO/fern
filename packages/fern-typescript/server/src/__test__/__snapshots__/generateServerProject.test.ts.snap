// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateServerProject posts 1`] = `
Object {
  "/.npmignore": "tsconfig.json",
  "/index.ts": "export * as model from \\"./model\\";
export * as services from \\"./services\\";
",
  "/model/commons/index.ts": "export * from \\"./types\\";
",
  "/model/commons/types/PostId.ts": "/**
 * An ID of a blog post
 */
export type PostId = string & {
    __PostId: void;
};

export const PostId = {
    of: (value: string): PostId => value as PostId
};
",
  "/model/commons/types/index.ts": "export * from \\"./PostId\\";
",
  "/model/index.ts": "export * as commons from \\"./commons\\";
export * as posts from \\"./posts\\";
",
  "/model/posts/errors/PostNotFoundError.ts": "import * as model from \\"../..\\";

export interface PostNotFoundError {
    id: model.commons.PostId;
}
",
  "/model/posts/errors/index.ts": "export * from \\"./PostNotFoundError\\";
",
  "/model/posts/index.ts": "export * from \\"./types\\";
export * from \\"./errors\\";
export * from \\"./service-types\\";
",
  "/model/posts/service-types/PostsService/CreatePostErrorBody.ts": "export type CreatePostErrorBody = never;

export declare namespace CreatePostErrorBody {
    export interface _Visitor<Result> {
        _unknown: () => Result;
    }
}

export const CreatePostErrorBody = {
    _visit: <Result>(_value: CreatePostErrorBody, visitor: CreatePostErrorBody._Visitor<Result>): Result => {
        return visitor._unknown();
    },

    _types: (): CreatePostErrorBody[\\"_error\\"][] => [],
} as const;
",
  "/model/posts/service-types/PostsService/CreatePostRequest.ts": "import * as model from \\"../../..\\";

export interface CreatePostRequest {
    type: model.posts.PostType;
    title: string;
    author: model.posts.Author;
    content: string;
}
",
  "/model/posts/service-types/PostsService/CreatePostResponse.ts": "import * as model from \\"../../..\\";

export type CreatePostResponse = CreatePostResponse.SuccessResponse | CreatePostResponse.ErrorResponse;

namespace CreatePostResponse {
    export interface SuccessResponse {
        ok: true;
        body: model.commons.PostId;
    }

    export interface ErrorResponse {
        ok: false;
        error: model.posts.CreatePostErrorBody;
    }
}
",
  "/model/posts/service-types/PostsService/DeletePostErrorBody.ts": "export type DeletePostErrorBody = never;

export declare namespace DeletePostErrorBody {
    export interface _Visitor<Result> {
        _unknown: () => Result;
    }
}

export const DeletePostErrorBody = {
    _visit: <Result>(_value: DeletePostErrorBody, visitor: DeletePostErrorBody._Visitor<Result>): Result => {
        return visitor._unknown();
    },

    _types: (): DeletePostErrorBody[\\"_error\\"][] => [],
} as const;
",
  "/model/posts/service-types/PostsService/DeletePostRequest.ts": "import * as model from \\"../../..\\";

export interface DeletePostRequest {
    postId: model.commons.PostId;
}
",
  "/model/posts/service-types/PostsService/DeletePostResponse.ts": "import * as model from \\"../../..\\";

export type DeletePostResponse = DeletePostResponse.SuccessResponse | DeletePostResponse.ErrorResponse;

namespace DeletePostResponse {
    export interface SuccessResponse {
        ok: true;
    }

    export interface ErrorResponse {
        ok: false;
        error: model.posts.DeletePostErrorBody;
    }
}
",
  "/model/posts/service-types/PostsService/GetPostErrorBody.ts": "import { posts } from \\"../../..\\";
import * as uuid from \\"uuid\\";

export type GetPostErrorBody =
    | GetPostErrorBody.PostNotFoundError;

export declare namespace GetPostErrorBody {
    interface PostNotFoundError extends posts.PostNotFoundError {
        _error: \\"PostNotFoundError\\";
        _errorInstanceId: string;
    }

    export interface _Visitor<Result> {
        PostNotFoundError: (value: posts.PostNotFoundError) => Result;
        _unknown: () => Result;
    }
}

export const GetPostErrorBody = {
    PostNotFoundError: (value: posts.PostNotFoundError): GetPostErrorBody.PostNotFoundError => ({
        ...value,
        _error: \\"PostNotFoundError\\",
        _errorInstanceId: uuid.v4()
    }),

    _visit: <Result>(value: GetPostErrorBody, visitor: GetPostErrorBody._Visitor<Result>): Result => {
        switch (value._error) {
            case \\"PostNotFoundError\\": return visitor.PostNotFoundError(value);
            default: return visitor._unknown();
        }
    },

    _types: (): GetPostErrorBody[\\"_error\\"][] => [\\"PostNotFoundError\\"],
} as const;
",
  "/model/posts/service-types/PostsService/GetPostRequest.ts": "import * as model from \\"../../..\\";

export interface GetPostRequest {
    postId: model.commons.PostId;
    page: number | null | undefined;
    otherParam: number | null | undefined;
}
",
  "/model/posts/service-types/PostsService/GetPostResponse.ts": "import * as model from \\"../../..\\";

export type GetPostResponse = GetPostResponse.SuccessResponse | GetPostResponse.ErrorResponse;

namespace GetPostResponse {
    export interface SuccessResponse {
        ok: true;
        body: model.posts.Post;
    }

    export interface ErrorResponse {
        ok: false;
        error: model.posts.GetPostErrorBody;
    }
}
",
  "/model/posts/service-types/PostsService/GetPostV2ErrorBody.ts": "import { posts } from \\"../../..\\";
import * as uuid from \\"uuid\\";

export type GetPostV2ErrorBody =
    | GetPostV2ErrorBody.PostNotFoundError;

export declare namespace GetPostV2ErrorBody {
    interface PostNotFoundError extends posts.PostNotFoundError {
        _error: \\"PostNotFoundError\\";
        _errorInstanceId: string;
    }

    export interface _Visitor<Result> {
        PostNotFoundError: (value: posts.PostNotFoundError) => Result;
        _unknown: () => Result;
    }
}

export const GetPostV2ErrorBody = {
    PostNotFoundError: (value: posts.PostNotFoundError): GetPostV2ErrorBody.PostNotFoundError => ({
        ...value,
        _error: \\"PostNotFoundError\\",
        _errorInstanceId: uuid.v4()
    }),

    _visit: <Result>(value: GetPostV2ErrorBody, visitor: GetPostV2ErrorBody._Visitor<Result>): Result => {
        switch (value._error) {
            case \\"PostNotFoundError\\": return visitor.PostNotFoundError(value);
            default: return visitor._unknown();
        }
    },

    _types: (): GetPostV2ErrorBody[\\"_error\\"][] => [\\"PostNotFoundError\\"],
} as const;
",
  "/model/posts/service-types/PostsService/GetPostV2Response.ts": "import * as model from \\"../../..\\";

export type GetPostV2Response = GetPostV2Response.SuccessResponse | GetPostV2Response.ErrorResponse;

namespace GetPostV2Response {
    export interface SuccessResponse {
        ok: true;
        body: model.posts.Post;
    }

    export interface ErrorResponse {
        ok: false;
        error: model.posts.GetPostV2ErrorBody;
    }
}
",
  "/model/posts/service-types/PostsService/index.ts": "export * from \\"./CreatePostRequest\\";
export * from \\"./CreatePostErrorBody\\";
export * from \\"./CreatePostResponse\\";
export * from \\"./GetPostRequest\\";
export * from \\"./GetPostErrorBody\\";
export * from \\"./GetPostResponse\\";
export * from \\"./GetPostV2ErrorBody\\";
export * from \\"./GetPostV2Response\\";
export * from \\"./DeletePostRequest\\";
export * from \\"./DeletePostErrorBody\\";
export * from \\"./DeletePostResponse\\";
",
  "/model/posts/service-types/index.ts": "export * from \\"./PostsService\\";
",
  "/model/posts/types/Author.ts": "export type Author =
    | Author.Anonymous
    | Author.Named;

export declare namespace Author {
    interface Anonymous {
        _type: \\"anonymous\\";
    }

    interface Named {
        _type: \\"named\\";
        named: string;
    }

    export interface _Visitor<Result> {
        anonymous: () => Result;
        named: (value: string) => Result;
        _unknown: () => Result;
    }
}

export const Author = {
    anonymous: (): Author.Anonymous => ({
        _type: \\"anonymous\\"
    }),

    named: (value: string): Author.Named => ({
        named: value,
        _type: \\"named\\"
    }),

    _visit: <Result>(value: Author, visitor: Author._Visitor<Result>): Result => {
        switch (value._type) {
            case \\"anonymous\\": return visitor.anonymous();
            case \\"named\\": return visitor.named(value.named);
            default: return visitor._unknown();
        }
    },

    _types: (): Author[\\"_type\\"][] => [\\"anonymous\\", \\"named\\"],
} as const;
",
  "/model/posts/types/Post.ts": "import * as model from \\"../..\\";

/**
 * A blog post
 */
export interface Post {
    id: model.commons.PostId;
    type: model.posts.PostType;
    title: string;
    author: model.posts.Author;
    content: string;
    relatedPosts: model.posts.Post[];
}
",
  "/model/posts/types/PostType.ts": "export type PostType =
    | \\"LONG\\"
    | \\"SHORT\\";

export const PostType = {
    Long: \\"LONG\\",
    Short: \\"SHORT\\",

    _visit: <Result>(value: PostType, visitor: PostType._Visitor<Result>): Result => {
        switch (value) {
            case PostType.Long: return visitor.long();
            case PostType.Short: return visitor.short();
            default: return visitor._unknown();
        }
    },

    _values: (): PostType[] => [PostType.Long, PostType.Short],
} as const;

export declare namespace PostType {
    type Long = \\"LONG\\";
    type Short = \\"SHORT\\";

    export interface _Visitor<Result> {
        long: () => Result;
        short: () => Result;
        _unknown: () => Result;
    }
}
",
  "/model/posts/types/index.ts": "export * from \\"./Post\\";
export * from \\"./PostType\\";
export * from \\"./Author\\";
",
  "/package.json": "{
    \\"name\\": \\"posts\\",
    \\"version\\": \\"0.0.0\\",
    \\"main\\": \\"./index.js\\",
    \\"types\\": \\"./index.d.ts\\",
    \\"scripts\\": {
        \\"build\\": \\"esbuild $(find . -name '*.ts') --format=cjs --outdir=. && tsc\\"
    },
    \\"dependencies\\": {
        \\"uuid\\": \\"8.3.2\\",
        \\"@types/uuid\\": \\"8.3.4\\",
        \\"@fern-typescript/service-utils\\": \\"0.0.100\\",
        \\"express\\": \\"4.18.1\\",
        \\"@types/express\\": \\"4.17.13\\"
    },
    \\"peerDependencies\\": {
        \\"express\\": \\"*\\",
        \\"@types/express\\": \\"*\\"
    },
    \\"devDependencies\\": {
        \\"@types/node\\": \\"17.0.33\\",
        \\"esbuild\\": \\"0.14.47\\",
        \\"typescript\\": \\"4.6.4\\"
    }
}",
  "/services/index.ts": "export * as posts from \\"./posts\\";
",
  "/services/posts/PostsService.ts": "import * as model from \\"../../model\\";
import { MaybePromise } from \\"@fern-typescript/service-utils\\";
import express, { Express } from \\"express\\";

export interface PostsService {
    createPost(request: model.posts.CreatePostRequest): MaybePromise<model.posts.CreatePostResponse>;
    getPost(request: model.posts.GetPostRequest): MaybePromise<model.posts.GetPostResponse>;
    getPostV2(request: model.commons.PostId): MaybePromise<model.posts.GetPostV2Response>;
    deletePost(request: model.posts.DeletePostRequest): MaybePromise<model.posts.DeletePostResponse>;
}

export const PostsService = {
    expressMiddleware: (impl: PostsService): Express => {
        const app = express();
        app.use(express.json());
        app.post(\\"/posts/\\", async (request, response) => {
            const result = await impl.createPost(request.body);
            if (result.ok) {
                response.send(result.body);
            }
            else {
                response.sendStatus(500);
            }
        });
        app.get(\\"/posts/:postId\\", async (request, response) => {
            const result = await impl.getPost({
                postId: model.commons.PostId.of(request.params.postId),
                page: request.query.page != null ? Number(request.query.page) : undefined,
                otherParam: request.query.otherParam != null ? Number(request.query.otherParam) : undefined
            });
            if (result.ok) {
                response.send(result.body);
            }
            else {
                const statusCode = model.posts.GetPostErrorBody._visit(result.error, {
                    PostNotFoundError: () => 404,
                    _unknown: () => 500
                });
                response.status(statusCode).send(result.error);
            }
        });
        app.get(\\"/posts/get\\", async (request, response) => {
            const result = await impl.getPostV2(request.body);
            if (result.ok) {
                response.send(result.body);
            }
            else {
                const statusCode = model.posts.GetPostV2ErrorBody._visit(result.error, {
                    PostNotFoundError: () => 404,
                    _unknown: () => 500
                });
                response.status(statusCode).send(result.error);
            }
        });
        app.delete(\\"/posts/:postId\\", async (request, response) => {
            const result = await impl.deletePost({
                postId: model.commons.PostId.of(request.params.postId)
            });
            if (result.ok) {
                response.end();
            }
            else {
                response.sendStatus(500);
            }
        });
        return app;
    }
};
",
  "/services/posts/index.ts": "export * from \\"./PostsService\\";
",
  "/tsconfig.json": "{
    \\"compilerOptions\\": {
        \\"strict\\": true,
        \\"skipLibCheck\\": true,
        \\"declaration\\": true,
        \\"sourceMap\\": true,
        \\"emitDeclarationOnly\\": true,
        \\"esModuleInterop\\": true
    }
}",
}
`;
